README.md
# Welcome to Paypal ruby library

This library is here to aid with integrating Paypal payments into ruby on rails
applications or similar. To set this up you will need to log into your paypal
business account and tell paypal where to send the IPN ( Instant payment notifications ).

# Download

* Preferred method of installation is using rubygems. gem install JonathanTron-paypal
* Alternatively you can get the source code at http://github.com/JonathanTron/paypal

# Requirements

* Ruby 1.8.2 (may work with previous versions) With OpenSSL support compiled in.
* Valid paypal business account.
* (optional) The money library from http://dist.leetsoft.com/api/money

# Installation

1. sudo gem install JonathanTron-paypal --source=http://gems.github.com

2. Require the library

        require "paypal"

  2.1. If you're using Rails add :

        require "paypal/rails"

3. Create a paypal_ipn ( or similar ) action like the one in the "Example action" appendix.


## -- - TODO : REWRITE BELOW THIS POINT - --

Within the new payment controller you can now create pages from which users can be sent to paypal. You always have to sent users to paypal using a HTTP Post so a standard link won't work (well OK but you need some javascript for that). The +Paypal::Helper+ namespace has some examples of how such a forward page may look.

# Testing the integration

Under https://developer.paypal.com/ you can signup for a paypal developer account.
This allows you to set up "sandboxed" accounts which work and act like real accounts
with the difference that no money is exchanged. Its a good idea to sign up for a
sandbox account to use while the application is running in development mode.


# Example rails controller

    class BackendController < ApplicationController

      # Simplification, please write better code then this...
      def paypal_ipn
       notify # Paypal::Notification.new(request.raw_post)

       if notify.acknowledge
         order # Order.find(notify.item_id)
         order.success # (notify.complete? and order.total # notify.amount) ? 'success' : 'failure'
         order.save
       end

       render :nothing #> true
      end
    end

# Example paypal forward page

   <%# paypal_form_tag %>
     <%# paypal_setup "Item 500", Money.us_dollar(50000), "bob@bigbusiness.com", :notify_url #> url_for(:only_path #> false, :action #> 'paypal_ipn') %>

     Please press here to pay $500US using paypal. <br/>
     <%# submit_tag "Go to paypal >>" %>

   </form>

   or, with the same results, the block version:

   <% paypal_form_tag do %>
     <%# paypal_setup "Item 500", Money.us_dollar(50000), "bob@bigbusiness.com", :notify_url #> url_for(:only_path #> false, :action #> 'paypal_ipn') %>

     Please press here to pay $500US using paypal. <br/>
     <%# submit_tag "Go to paypal >>" %>

   <% end %>

# Using encrypted form data

Paypal supports encrypted form data to prevent tampering by third parties.
You must have a verified paypal account to use this functionality.

1) Create a private key for yourself

    openssl genrsa -out business_key.pem 1024

2) Create a public certificate to share with Paypal

    openssl req -new -key business_key.pem -x509 -days 3650 -out business_cert.pem

3) Upload the public certificate to Paypal (under Profile -> Encrypted Payment Settings -> Your Public Certificates -> Add),
and note the "Cert ID" that Paypal shows for the certificate.

4) Update your controller to include the details for your key and certificate.

    @business_key # File::read("business_key.pem")
    @business_cert # File::read("business_cert.pem")
    @business_certid # "certid from paypal"

5) Update your views to populate the :business_key, :business_cert and :business_certid options in 'paypal_setup' - the rest of the signature is the same.

6) When you're ready to go live, download the production Paypal certificate and override the default certificate.

    Paypal::Notification.paypal_cert # File::read("paypal_cert.pem")

7) Finally, add the following line to your environment.rb or inside an initializer:

    Paypal::Notification.ipn_url # "https://www.paypal.com/cgi-bin/webscr"

# Troubleshooting

uninitalized constant Paypal - Make sure your ruby has openssl support

CHANGELOG.md
= 3.0.0 (git)

    Update spec task to use the new RSpec module name
    Remove constant/class variable usage and use Thread specific configurations
    Start spec on Paypal::Helpers::Common.paypal_setup
    Fix specs for Paypal::Config
    Upgrade dev dependencies to use Rspec >= 2.0.0.a
    Introduce a Paypal::Config module to handle ipn/certificates (Jonathan Tron)
    Allow access to params as method (via method_missing) (Jonathan Tron)
    Rework Paypal::Notification :
      Remove methods to get params informations (CAUTION !!! THIS BREAK OLD API) (Jonathan Tron)
      Add methods for all possible statuses and rename #complete? as completed? to follow Paypal naming (Jonathan Tron)
    Use Rack::Utils#parse_query in Paypal::Notification for query parsing (Jonathan Tron)
    Add Spec for notification (Jonathan Tron)
    Add paypal/rails.rb to trigger helper inclusion in view (Jonathan Tron)
    Separate actual helpers in two files (paypal/helpers/common.rb and paypal/helpers/rails.rb) (Jonathan Tron)
    Move files around (Jonathan Tron)
    Update the README and switch to MarkDown (Jonathan Tron)
    Add a CHANGELOG file and switch to new format (Jonathan Tron)
    Remove init.rb (Jonathan Tron)
    Configure gem generation with jelewer (Jonathan Tron)

= 2008-10-16 -- 2.0.2

    NEW: Added block style support for paypal_form_tag (paypal_form_tag do blah.. blah... end)
    DEL: Removed patch for 2.0.0, no longer seems suitable.
    NEW: Added testing for the paypal_form_tag block style.
    NEW: Added a sample (actual from paypal sandbox) PayPal server response for IPN, to aid in testing.

= 2008-10-15 -- 2.0.1

    CHG: Modified README.
    FIX: Moved patch to own directory, was being caught by git-hub as gem's README
    NEW: Added patch for currently installed paypal-2.0.0 gem to apply directly on gems directory.
    NEW: Added relevant test statements.
    FIX: removed duplicate 'invoice' method in lib/notification.rb
    NEW: added correct 'custom' method to lib/notification.rb
    NEW: added pending_reason, reason_code, memo, payment_type, exchange_rate methods to lib/notification.rb

= 2006-04-20 -- 2.0.0

    Uses paypal extended syntax. The plugin can now submit shipping and billing addresses using the paypal_address helper.

= 2006-04-20 -- 1.7.0 

    Now a rails plugin

= 2006-02-10 -- 1.5.1

    added complete list of valid paypal options (Paul Hart)

= 2006-02-02 -- 1.5.0

    Now report an error when invalid option is passed to paypal_setup
    Had to rename parameters cancel_url to cancel_return and return_url to return, please update your app
    Improved the test coverage strategy for helper tests
    Added support for encrypted form data (Paul Hart)

= 2005-09-16 -- 0.9.6

    Added readme note about the openssl requirement

= 2005-07-26 -- 0.9.5

    Added tax to the helper parameters
    fixed bug when money class was used to pass in amount. Cents were always 00 (doh!)
    Added invoice and custom optional parameters
    Added charset = utf-8 to all paypal posts
    Wrongly used undefined_quanitity parameter in 0.9.1, this caused users to be prompted for the quanitity on the paypal checkout page... fixed

= 2005-07-22 -- 0.9.1

    support for cancel_url as well as notify_url. This means you can now set the IPN callback address from the paypal_setup method and 
  you don't have to do that in the paypal admin interface!
    Removed the actual form tag from the paypal_setup generated code to conform better with docs 


spec/config_spec.rb
require "spec_helper"

describe Paypal::Config do
  after do
    Thread.current.keys.each do |key|
      Thread.current[key] = nil unless key =~ /^__.+__$/
    end
    Paypal.class_eval do
      remove_const :Config if const_defined?(:Config)
    end
    load "paypal/config.rb"
  end
  
  describe "ipn_urls" do
    it "should define url for :production" do
      Paypal::Config.ipn_urls[:production].should eql("https://www.paypal.com/cgi-bin/webscr")
    end
    it "should define url for :sandbox" do
      Paypal::Config.ipn_urls[:sandbox].should eql("https://www.sandbox.paypal.com/cgi-bin/webscr")
    end
    it "should allow changing urls" do
      url = "https://myownsandbox.com"
      Paypal::Config.ipn_urls[:sandbox] = url
      Paypal::Config.ipn_urls[:sandbox].should eql(url)
    end
  end

  describe "mode" do
    it "should be :sandbox by default" do
      Paypal::Config.mode.should eql(:sandbox)
    end
    
    it "should accept :sandbox" do
      lambda {
        Paypal::Config.mode = :sandbox
      }.should_not raise_error
    end

    it "should accept :production" do
      lambda {
        Paypal::Config.mode = :production
      }.should_not raise_error
    end

    it "should not accept something different than :production or :sandbox" do
      lambda {
        Paypal::Config.mode = :demo
      }.should raise_error(ArgumentError)
    end
  end
  
  describe "when setting mode to sandbox" do
    before do
      Paypal::Config.mode = :sandbox
    end
    
    it "should be in :sandbox mode" do
      Paypal::Config.mode.should eql(:sandbox)
    end

    it "should have ipn url for sandbox" do
      Paypal::Config.ipn_url.should eql("https://www.sandbox.paypal.com/cgi-bin/webscr")
    end
    
    it "should have the ipn validation path for sandbox" do
      Paypal::Config.ipn_validation_path.should eql("/cgi-bin/webscr?cmd=_notify-validate")
    end

    it "should have the ipn validation url for sandbox" do
      Paypal::Config.ipn_validation_url.should eql("https://www.sandbox.paypal.com/cgi-bin/webscr?cmd=_notify-validate")
    end
    
    it "should have paypal cert for sandbox" do
      Paypal::Config.paypal_cert.should eql(
"-----BEGIN CERTIFICATE-----
MIIDoTCCAwqgAwIBAgIBADANBgkqhkiG9w0BAQUFADCBmDELMAkGA1UEBhMCVVMx
EzARBgNVBAgTCkNhbGlmb3JuaWExETAPBgNVBAcTCFNhbiBKb3NlMRUwEwYDVQQK
EwxQYXlQYWwsIEluYy4xFjAUBgNVBAsUDXNhbmRib3hfY2VydHMxFDASBgNVBAMU
C3NhbmRib3hfYXBpMRwwGgYJKoZIhvcNAQkBFg1yZUBwYXlwYWwuY29tMB4XDTA0
MDQxOTA3MDI1NFoXDTM1MDQxOTA3MDI1NFowgZgxCzAJBgNVBAYTAlVTMRMwEQYD
VQQIEwpDYWxpZm9ybmlhMREwDwYDVQQHEwhTYW4gSm9zZTEVMBMGA1UEChMMUGF5
UGFsLCBJbmMuMRYwFAYDVQQLFA1zYW5kYm94X2NlcnRzMRQwEgYDVQQDFAtzYW5k
Ym94X2FwaTEcMBoGCSqGSIb3DQEJARYNcmVAcGF5cGFsLmNvbTCBnzANBgkqhkiG
9w0BAQEFAAOBjQAwgYkCgYEAt5bjv/0N0qN3TiBL+1+L/EjpO1jeqPaJC1fDi+cC
6t6tTbQ55Od4poT8xjSzNH5S48iHdZh0C7EqfE1MPCc2coJqCSpDqxmOrO+9QXsj
HWAnx6sb6foHHpsPm7WgQyUmDsNwTWT3OGR398ERmBzzcoL5owf3zBSpRP0NlTWo
nPMCAwEAAaOB+DCB9TAdBgNVHQ4EFgQUgy4i2asqiC1rp5Ms81Dx8nfVqdIwgcUG
A1UdIwSBvTCBuoAUgy4i2asqiC1rp5Ms81Dx8nfVqdKhgZ6kgZswgZgxCzAJBgNV
BAYTAlVTMRMwEQYDVQQIEwpDYWxpZm9ybmlhMREwDwYDVQQHEwhTYW4gSm9zZTEV
MBMGA1UEChMMUGF5UGFsLCBJbmMuMRYwFAYDVQQLFA1zYW5kYm94X2NlcnRzMRQw
EgYDVQQDFAtzYW5kYm94X2FwaTEcMBoGCSqGSIb3DQEJARYNcmVAcGF5cGFsLmNv
bYIBADAMBgNVHRMEBTADAQH/MA0GCSqGSIb3DQEBBQUAA4GBAFc288DYGX+GX2+W
P/dwdXwficf+rlG+0V9GBPJZYKZJQ069W/ZRkUuWFQ+Opd2yhPpneGezmw3aU222
CGrdKhOrBJRRcpoO3FjHHmXWkqgbQqDWdG7S+/l8n1QfDPp+jpULOrcnGEUY41Im
jZJTylbJQ1b5PBBjGiP0PpK48cdF
-----END CERTIFICATE-----")
    end
    
    it "should allow setting a new sandbox cert" do
      lambda {
        Paypal::Config.paypal_sandbox_cert = "TEST"
      }.should_not raise_error
      Paypal::Config.paypal_cert.should eql("TEST")
    end
  end
  
  describe "when setting mode to :production" do
    before do
      Paypal::Config.mode = :production
    end
    
    it "should be in :production mode" do
      Paypal::Config.mode.should eql(:production)
    end

    it "should have ipn url for production" do
      Paypal::Config.ipn_url.should eql("https://www.paypal.com/cgi-bin/webscr")
    end
    
    it "should have the ipn validation path for production" do
      Paypal::Config.ipn_validation_path.should eql("/cgi-bin/webscr?cmd=_notify-validate")
    end

    it "should have the ipn validation url for production" do
      Paypal::Config.ipn_validation_url.should eql("https://www.paypal.com/cgi-bin/webscr?cmd=_notify-validate")
    end

    describe "when paypal_production_cert was not set" do
      before do
        @old_cert, Paypal::Config.paypal_production_cert = Paypal::Config.paypal_production_cert, nil
      end
      
      after do
        Paypal::Config.paypal_production_cert = @old_cert
      end
      
      it "should raise an error" do
        lambda {
          Paypal::Config.paypal_cert
        }.should raise_error(StandardError)
      end
    end

    describe "when paypal_production_cert was set" do
      before do
        @old_cert, Paypal::Config.paypal_production_cert = Paypal::Config.paypal_production_cert, "TEST"
      end
      
      after do
        Paypal::Config.paypal_production_cert = @old_cert
      end
      
      it "should not raise an error" do
        lambda {
          Paypal::Config.paypal_cert
        }.should_not raise_error(StandardError)
      end

      it "should use the paypal production certificate" do
        Paypal::Config.paypal_cert.should eql("TEST")
      end
    end

    it "should allow setting a new production cert" do
      lambda {
        Paypal::Config.paypal_production_cert = "TEST"
      }.should_not raise_error
      Paypal::Config.paypal_cert.should eql("TEST")
    end
  end
  
  describe "in multithreaded environment" do
    it "should not mix different thread values" do
      t = Thread.new { 
        Paypal::Config.mode = :production
        Paypal::Config.mode.should eql(:production)
      }
      t.join
      Paypal::Config.mode.should eql(:sandbox)
    end
  end
end

spec/spec_helper.rb
begin
  # Try to require the preresolved locked set of gems.
  require ::File.expand_path('../.bundle/environment', __FILE__)
rescue LoadError
  # Fall back on doing an unlocked resolve at runtime.
  require "rubygems"
  require "bundler"
  Bundler.setup
end

Bundler.require :default

require "paypal"

# Do not allow connection to non registered URLs, so we can catch if specifics were called
FakeWeb.allow_net_connect = false

Rspec::Matchers.define :have_css do |css|
  match do |text|
    html = Nokogiri::HTML(text)
    !html.css(css).empty?
  end
 
  failure_message_for_should do |text|
    "expected to find css expression #{css} in:\n#{text}"
  end
 
  failure_message_for_should_not do |text|
    "expected not to find css expression #{css} in:\n#{text}"
  end

  description do
    "have css in #{expected}"
  end
end

Rspec.configure do |c|
  c.mock_framework = :rspec
  c.color_enabled = true
end

spec/notification_spec.rb
require File.expand_path(File.dirname(__FILE__) + '/spec_helper')

describe Paypal::Notification do
  before do
    Paypal::Config.mode = :sandbox
  end
  
  def valid_http_raw_data
    "mc_gross=500.00&address_status=confirmed&payer_id=EVMXCLDZJV77Q&tax=0.00&address_street=164+Waverley+Street&payment_date=15%3A23%3A54+Apr+15%2C+2005+PDT&payment_status=Completed&address_zip=K2P0V6&first_name=Tobias&mc_fee=15.05&address_country_code=CA&address_name=Tobias+Luetke&notify_version=1.7&custom=cusdata&payer_status=unverified&business=tobi%40leetsoft.com&address_country=Canada&address_city=Ottawa&quantity=1&payer_email=tobi%40snowdevil.ca&verify_sign=AEt48rmhLYtkZ9VzOGAtwL7rTGxUAoLNsuf7UewmX7UGvcyC3wfUmzJP&txn_id=6G996328CK404320L&payment_type=instant&last_name=Luetke&address_state=Ontario&receiver_email=tobi%40leetsoft.com&payment_fee=&receiver_id=UQ8PDYXJZQD9Y&txn_type=web_accept&item_name=Store+Purchase&mc_currency=CAD&item_number=&test_ipn=1&payment_gross=&shipping=0.00&invoice=myinvoice&pending_reason=mypending_reason&reason_code=myreason_code&memo=mymemo&payment_type=mypayment_type&exchange_rate=myexchange_rate"
  end

  def self.valid_parsed_raw_data
    {"payment_gross"=>"", "receiver_id"=>"UQ8PDYXJZQD9Y", "payer_email"=>"tobi@snowdevil.ca", "address_city"=>"Ottawa", "address_country"=>"Canada", "business"=>"tobi@leetsoft.com", "address_name"=>"Tobias Luetke", "payment_status"=>"Completed", "tax"=>"0.00", "reason_code"=>"myreason_code", "receiver_email"=>"tobi@leetsoft.com", "invoice"=>"myinvoice", "verify_sign"=>"AEt48rmhLYtkZ9VzOGAtwL7rTGxUAoLNsuf7UewmX7UGvcyC3wfUmzJP", "address_street"=>"164 Waverley Street", "memo"=>"mymemo", "mc_currency"=>"CAD", "txn_type"=>"web_accept", "quantity"=>"1", "address_zip"=>"K2P0V6", "pending_reason"=>"mypending_reason", "item_name"=>"Store Purchase", "txn_id"=>"6G996328CK404320L", "address_country_code"=>"CA", "payment_fee"=>"", "address_state"=>"Ontario", "payer_status"=>"unverified", "notify_version"=>"1.7", "shipping"=>"0.00", "mc_fee"=>"15.05", "payment_date"=>"15:23:54 Apr 15, 2005 PDT", "address_status"=>"confirmed", "test_ipn"=>"1", "payment_type"=>"mypayment_type", "first_name"=>"Tobias", "last_name"=>"Luetke", "payer_id"=>"EVMXCLDZJV77Q", "mc_gross"=>"500.00", "exchange_rate"=>"myexchange_rate", "item_number"=>"", "custom"=>"cusdata"}
  end
  
  describe "without data" do
    it "should raise Paypal::NoDataError" do
      lambda {
        Paypal::Notification.new(nil)
      }.should raise_error(Paypal::NoDataError)
    end
  end
  
  describe "with valid raw data" do
    before do
      @notification = Paypal::Notification.new(valid_http_raw_data)
    end

    it "should store raw data" do
      @notification.raw.should eql(valid_http_raw_data)
    end

    describe "#params" do
      it "should not be empty" do
        @notification.params.should_not be_empty
      end

      valid_parsed_raw_data.each do |key, value|
        it "should have #{key}=#{value}" do
          @notification.params[key].should eql(value)
        end
      end

      it "should have unescaped values" do
        @notification.params["payment_date"].should eql("15:23:54 Apr 15, 2005 PDT")
      end
      
      it "should include only the last value for duplicate key" do
        @notification.params["payment_type"].should eql("mypayment_type")
      end
    end

    it "should define method to access each query params" do
      self.class.valid_parsed_raw_data.each do |key, _|
        lambda {
          @notification.send(key.to_sym)
        }.should_not raise_error
      end
    end

    ["Canceled_Reversal",
    "Completed",
    "Denied",
    "Expired",
    "Failed",
    "Pending",
    "Processed",
    "Refunded",
    "Reversed",
    "Voided"].each do |status|
      describe "when transaction payment_status = '#{status}'" do
        it "should be #{status.downcase}" do
          old_status, @notification.params["payment_status"] = @notification.params["payment_status"], status
        
          @notification.send(:"#{status.downcase}?").should be_true
        
          @notification.params["payment_status"] = old_status
        end
      end
    end

    describe "#acknowledge" do
      before do
        @paypal_validation_url = "https://www.sandbox.paypal.com/cgi-bin/webscr?cmd=_notify-validate"
      end
      
      it "should send a post request to Paypal at #{@paypal_validation_url}" do
        FakeWeb.register_uri(:post, @paypal_validation_url, :body => 'VERIFIED')
        
        lambda { @notification.acknowledge }.should_not raise_error(FakeWeb::NetConnectNotAllowedError)

        FakeWeb.clean_registry
      end
      
      describe "when Paypal response is VERIFIED" do
        before do
          FakeWeb.register_uri(:post, @paypal_validation_url, :body => 'VERIFIED')
        end

        it "should return true" do
          @notification.acknowledge.should be(true)
        end

        after do
          FakeWeb.clean_registry
        end
      end

      describe "when Paypal response is INVALID" do
        before do
          FakeWeb.register_uri(:post, @paypal_validation_url, :body => 'INVALID')
        end

        it "should return false" do
          @notification.acknowledge.should be(false)
        end

        after do
          FakeWeb.clean_registry
        end
      end

      describe "when Paypal response is not a recognize value" do
        before do
          FakeWeb.register_uri(:post, @paypal_validation_url, :body => 'BAD_VALUE')
        end

        it "should raise StandardError" do
          lambda {
            @notification.acknowledge
          }.should raise_error(StandardError)
        end
        
        it "should include body response in StandardError" do
          begin
            @notification.acknowledge
          rescue StandardError => e
            e.message.should =~ /BAD_VALUE/
          end
        end

        after do
          FakeWeb.clean_registry
        end
      end
    end
  end
end

lib/paypal/config.rb
module Paypal
  module Config
    def self.ipn_urls
      Thread.current[:ipn_urls] ||= {
        :sandbox => "https://www.sandbox.paypal.com/cgi-bin/webscr",
        :production => "https://www.paypal.com/cgi-bin/webscr"
      }
    end

    def self.mode=(new_mode)
      raise ArgumentError.new("Paypal::Config.mode should be either :sandbox or :production (you tried to set it as : #{new_mode})") unless [:sandbox, :production].include?(new_mode.to_sym)
      Thread.current[:mode] = new_mode.to_sym
    end
    def self.mode
      Thread.current[:mode] ||= :sandbox
    end
    
    def self.ipn_url
      ipn_urls[mode]
    end
    
    def self.ipn_validation_path
      URI.parse(ipn_url).path + "?cmd=_notify-validate"
    end

    def self.ipn_validation_url
      "#{ipn_url}?cmd=_notify-validate"
    end

    def self.paypal_sandbox_cert=(new_cert)
      Thread.current[:paypal_sandbox_cert] = new_cert
    end
    def self.paypal_sandbox_cert
      Thread.current[:paypal_sandbox_cert] ||= File.read(File.join(File.dirname(__FILE__), 'certs', 'paypal_sandbox.pem'))
    end

    def self.paypal_production_cert=(new_cert)
      Thread.current[:paypal_production_cert] = new_cert
    end
    def self.paypal_production_cert
      Thread.current[:paypal_production_cert]
    end

    def self.paypal_cert
      case mode
      when :sandbox
        paypal_sandbox_cert
      when :production
        raise StandardError.new("You should set Paypal::Config.paypal_production_cert with your paypal production certificate") unless paypal_production_cert
        paypal_production_cert
      end
    end
  end
end

lib/paypal/helpers/common.rb
module Paypal
  # This is a collection of helpers which aid in the creation of paypal buttons
  #
  # Example:
  #
  #    <%= form_tag Paypal::Config.ipn_url %>
  #
  #      <%= paypal_setup "Item 500", Money.us_dollar(50000), "bob@bigbusiness.com" %>
  #      Please press here to pay $500US using paypal. <%= submit_tag %>
  #
  #    <% end_form_tag %>
  #
  # For this to work you have to include these methods as helpers in your rails application.
  # One way is to add "include Paypal::Helpers" in your application_helper.rb
  # See Paypal::Notification for information on how to catch payment events.
  module Helpers
    module Common
      # This helper creates the hidden form data which is needed for a paypal purchase.
      #
      # * <tt>item_number</tt> -- The first parameter is the item number. This is for your personal organization and can
      #   be arbitrary. Paypal will sent the item number back with the IPN so its a great place to
      #   store a user ID or a order ID or something like  this.
      #
      # * <tt>amount</tt> -- should be a parameter of type Money ( see http://leetsoft.com/api/money ) but can also
      #   be a string of type "50.00" for 50$. If you use the string syntax make sure you set the current
      #   currency as part of the options hash. The default is USD
      #
      # * <tt>business</tt> -- This is your paypal account name ( an email ). This needs to be a valid paypal business account.
      #
      # The last parameter is a options hash. You can set or override any Paypal-recognized parameter, including:
      #
      # * <tt>:cmd</tt> -- default is '_xclick' or '_xclick-subscriptions' when you use :subscription.
      # * <tt>:quantity</tt> -- default is '1'.
      # * <tt>:no_note</tt> -- default is '1'.
      # * <tt>:item_name</tt> -- default is 'Store purchase'. This is the name of the purchase which will be displayed
      #   on the paypal page.
      # * <tt>:no_shipping</tt> -- default is '1'. By default we tell paypal that no shipping is required. Usually
      #   the shipping address should be collected in our application, not by paypal.
      # * <tt>:currency</tt> -- default is 'USD'. If you provide a Money object, that will automatically override
      #   the value.
      # * <tt>:charset</tt> -- default is 'utf-8'.
      # * <tt>:notify_url</tt> -- If provided paypal will send its IPN notification once a
      #   purchase is made, canceled or any other status changes occur.
      # * <tt>:return</tt> -- If provided paypal will redirect a user back to this url after a
      #   successful purchase. Useful for a kind of thankyou page.
      # * <tt>:cancel_return</tt> -- If provided paypal will redirect a user back to this url when
      #   the user cancels the purchase.
      # * <tt>:tax</tt> -- the tax for the store purchase. Same format as the amount parameter but optional
      # * <tt>:invoice</tt> -- Unique invoice number. User will never see this. optional
      # * <tt>:custom</tt> -- Custom field. User will never see this. optional
      #
      # Dealing with subscriptions
      #
      # * <tt>:subscription</tt> -- Hash containing the subscription options. optional
      #   * <tt>:period</tt> -- One of :monthly, :yearly, :weekly or :daily
      #   * <tt>:length</tt> -- How often, based on :period. E.g. 6 :monthly?
      #   * <tt>:retry</tt> -- Default is false. Boolean for if paypal should retry failed payments.
      #   * <tt>:recurring</tt> -- Default is false. Boolean for if paypal should recur payment and end of period.
      #
      # Generating encrypted form data
      #
      # The helper also supports the generation of encrypted button data. Please see the README for more information
      # on the setup and prerequisite steps for using encrypted forms.
      #
      # The following options must all be provided (as strings) to encrypt the data:
      #
      # * <tt>:business_key</tt> -- The private key you have generated
      # * <tt>:business_cert</tt> -- The public certificate you have also uploaded to Paypal
      # * <tt>:business_certid</tt> -- The certificate ID that Paypal has assigned to your certificate.
      #
      # Examples:
      #
      #   <%= paypal_setup @order.id, Money.us_dollar(50000), "bob@bigbusiness.com" %>
      #   <%= paypal_setup @order.id, '50.00', "bob@bigbusiness.com", :currency => 'USD' %>
      #   <%= paypal_setup @order.id, '50.00', "bob@bigbusiness.com", :currency => 'USD', :notify_url => url_for(:only_path => false, :action => 'paypal_ipn') %>
      #   <%= paypal_setup @order.id, Money.ca_dollar(50000), "bob@bigbusiness.com", :item_name => 'Snowdevil shop purchase', :return_url => paypal_return_url, :cancel_url => paypal_cancel_url, :notify_url => paypal_ipn_url  %>
      #   <%= paypal_setup @order.id, Money.ca_dollar(50000), "bob@bigbusiness.com", :item_name => 'Snowdevil shop purchase', :return_url => paypal_return_url, :cancel_url => paypal_cancel_url, :business_key => @business_key, :business_cert => @business_cert, :business_certid => @business_certid  %>
      #
      def paypal_setup(item_number, amount, business, options = {})

        subscription = options.delete(:subscription)

        misses = (options.keys - valid_setup_options)
        raise ArgumentError, "Unknown option #{misses.inspect}" unless misses.empty?

        params = {
          :cmd => subscription ? '_ext-enter' : '_xclick',
          :redirect_cmd => subscription ? '_xclick-subscriptions' : nil,
          :quantity => 1,
          :business => business,
          :item_number => item_number,
          :item_name => 'Store purchase',
          :no_shipping => '1',
          :no_note => '1',
          :charset => 'utf-8'
        }.reject{|k,v| v.nil?}.merge(options)

        params[:currency_code] = amount.currency if amount.respond_to?(:currency)
        params[:currency_code] = params.delete(:currency) if params[:currency]
        params[:currency_code] ||= 'USD'

        # We accept both strings and money objects as amount
        amount = amount.cents.to_f / 100.0 if amount.respond_to?(:cents)
        amount = sprintf('%.2f', amount)

        if subscription.nil?
          params[:amount] = amount
        else
          params[:a3]  = amount
          params[:p3]  = subscription[:length]
          params[:sra] = subscription[:retry] == true ? 1 : 0
          params[:src] = subscription[:recurring] == true ? 1 : 0
          params[:t3]  = case subscription[:period]
            when :monthly; 'M'
            when :yearly;  'Y'
            when :weekly;  'W'
            when :daily;   'D'
          end
        end

        # same for tax
        tax = params[:tax]
        unless tax.nil?
          tax = tax.cents.to_f / 100.0 if tax.respond_to?(:cents)
          params[:tax] = sprintf("%.2f", tax)
        end

        # look for encryption parameters, save them outsite the parameter hash.
        business_key = params.delete(:business_key)
        business_cert = params.delete(:business_cert)
        business_certid = params.delete(:business_certid)

        # Build the form
        buttons = []
        # Only attempt an encrypted form if we have all the required fields.
        if business_key and business_cert and business_certid
          require 'openssl'

          # Convert the key and certificates into OpenSSL-friendly objects.
          paypal_cert = OpenSSL::X509::Certificate.new(Paypal::Config.paypal_cert)
          business_key = OpenSSL::PKey::RSA.new(business_key)
          business_cert = OpenSSL::X509::Certificate.new(business_cert)
          # Put the certificate ID back into the parameter hash the way Paypal wants it.
          params[:cert_id] = business_certid

          # Prepare a string of data for encryption
          data = ""
          params.each_pair {|k,v| data << "#{k}=#{v}\n"}

          # Sign the data with our key/certificate pair
          signed = OpenSSL::PKCS7::sign(business_cert, business_key, data, [], OpenSSL::PKCS7::BINARY)
          # Encrypt the signed data with Paypal's public certificate.
          encrypted = OpenSSL::PKCS7::encrypt([paypal_cert], signed.to_der, OpenSSL::Cipher::Cipher::new("DES3"), OpenSSL::PKCS7::BINARY).to_s.gsub("\n", "")

          # The command for encrypted forms is always '_s-xclick'; the real command is in the encrypted data.
          buttons << %Q{<input type="hidden" name="cmd" value="_s-xclick" />}
          buttons << %Q{<input type="hidden" name="encrypted" value="#{encrypted}" />}
        else
          # Just emit all the parameters that we have as hidden fields.
          # Note that the sorting isn't really needed, but it makes testing a lot easier for now.
          params.each do |key, value|
            buttons << %Q{<input type="hidden" name="#{key}" value="#{value}" />} unless value.nil?
          end
        end
        buttons.join("\n")
      end

      # Pass an address to paypal so that all signup forms can be prefilled
      #
      # * <tt>email</tt> -- Customer's email address
      # * <tt>first_name</tt> --  Customer's first name. Must be alpha-numeric, with a 32 character limit
      # * <tt>last_name</tt> -- Customer's last name. Must be alpha-numeric, with a 64 character limit
      # * <tt>address1</tt> --  First line of customer's address. Must be alpha-numeric, with a 100 character limit
      # * <tt>address2</tt> --  Second line of customer's address. Must be alpha-numeric, with a 100 character limit
      # * <tt>city</tt> --  City of customer's address. Must be alpha-numeric, with a 100 character limit
      # * <tt>state</tt> -- State of customer's address. Must be official 2 letter abbreviation
      # * <tt>zip</tt> -- Zip code of customer's address
      # * <tt>night_phone_a</tt> -- Area code of customer's night telephone number
      # * <tt>night_phone_b</tt> -- First three digits of customer's night telephone number
      # * <tt>day_phone_a</tt> -- Area code of customer's daytime telephone number
      # * <tt>day_phone_b</tt> -- First three digits of customer's daytime telephon
      def paypal_address(options = {})
        options.collect do |key, value|
          %Q{<input type="hidden" name="#{key}" value="#{value}" />}
        end.join("\n")
      end

    private

      # See https://www.paypal.com/IntegrationCenter/ic_std-variable-reference.html for details on the following options.
      def valid_setup_options
        [
        # Generic Options
        :cmd,
        # IPN Support
        :notify_url,
        # Item Information
        :item_name,
        :quantity,
        :undefined_quantity,
        :on0,
        :os0,
        :on1,
        :os1,
        # Display Information
        :add,
        :cancel_return,
        :cbt,
        :cn,
        :cpp_header_image,
        :cpp_headerback_color,
        :cpp_headerborder_color,
        :cpp_payflow_color,
        :cs,
        :display,
        :image_url,
        :no_note,
        :no_shipping,
        :page_style,
        :return,
        :rm,
        # Transaction Information
        :address_override,
        :currency,
        :currency_code,
        :custom,
        :handling,
        :invoice,
        :redirect_cmd,
        :shipping,
        :tax,
        :tax_cart,
        # Shopping Cart Options
        :amount,
        :business,
        :handling_cart,
        :paymentaction,
        :rupload,
        :charset,
        :upload,
        # Prepopulating PayPal FORMs or Address Overriding
        :address1,
        :address2,
        :city,
        :country,
        :email,
        :first_name,
        :last_name,
        :lc,
        :night_phone_a,
        :night_phone_b,
        :night_phone_c,
        :state,
        :zip,
        # Prepopulating Business Account Sign-up
        :business_address1,
        :business_address2,
        :business_city,
        :business_state,
        :business_country,
        :business_cs_email,
        :business_cs_phone_a,
        :business_cs_phone_b,
        :business_cs_phone_c,
        :business_url,
        :business_night_phone_a,
        :business_night_phone_b,
        :business_night_phone_c,
        # End of list from https://www.paypal.com/IntegrationCenter/ic_std-variable-reference.html
        # The following items are known to exist but are not yet on the above page.
        :business_zip,
        :day_phone_a,
        :day_phone_b,
        :day_phone_c,
        # Subscription Options
        :a1, # Trial Amount 1
        :p1, # Trial Period 1
        :t1, # Trial Period 1 Units (D=days, W=weeks, M=months, Y=years)
        :a2, # Trial Amount 2
        :p2, # Trial Period 2
        :t2, # Trial Period 2 Units
        :a3, # Regular Subscription Amount
        :p3, # Regular Subscription Period
        :t3, # Regular Subscription Period Units
        :src, # Recurring Payments? (1=yes, default=0)
        :sra, # Reattempt Transaction on Failure? (1=yes, default=0)
        :srt, # Recurring Times (number of renewals before auto-cancel, default=forever)
        :usr_manage, # Username and Password Generator? (1=yes, default=0)
        :modify, # Modification Behaviour (0=new subs only, 1=new or modify, 2=modify existing only, default=0)
        # Encryption Options - used internally only.
        :business_key, # Your private key
        :business_cert, # Your public certificate
        :business_certid, # Your public certificate ID (from Paypal)
        # Other
        :bn
        ]
      end
    end
  end
end
